feat(music): add album and artist methods

jeremyhofer opened the pull request:
This PR introduces two new methods to the `music` module. `albumName` and `artist`. 

Data was sourced from various resources from a number of resources online, including:
- The dataset from when the `songName` method was added in https://github.com/faker-js/faker/pull/996: https://manghammath.com/80s%20Hits/The%20Top%201000%20Songs%20of%20AllTime.pdf
- Rolling Stone 500 Greatest Albums: https://www.rollingstone.com/music/music-lists/best-albums-of-all-time-1062063/arcade-fire-%ef%bb%bffuneral-1062733/
- Chart Masters compiled data for Artists and Albums: https://chartmasters.org/

The `@since` version on the methods was set to `8.5.0`, as it looks like the `8.4.0` tag has been created.
<!-- Please run `pnpm run preflight` before opening a Pull Request to ensure that your code fulfills the minimal requirements for our project. -->

<!-- Please first read https://github.com/faker-js/faker/blob/next/CONTRIBUTING.md -->

<!-- Help us by writing a correct PR title following this guide: https://github.com/faker-js/faker/blob/next/CONTRIBUTING.md#committing -->

jeremyhofer committed:
feat(music): add albumName and artist methods add additional genres and songs

matthewmayer reviewed (changes requested):
1. We normally limit to 1000 entries per file
2. Suggest splitting into two PRs, one for adding data for existing methods (since that can be done in a patch release) and one for proposing new methods
3. Some of the genres look a bit awkward like "Holiday: Other". Maybe remove ones like that.

jeremyhofer commented:
@matthewmayer thanks for the feedback!

I pulled the genre and song name additions out of this PR. Will raise a PR to add and clean up the additional genres. Given the song name set is near 1000 entries I'm discarding all my changes there.

I updated the added `album_name` and `artist` sets down to 1000 entries each.

jeremyhofer committed:
properly sort album and artist lists. update snapshot

matthewmayer commented on src/modules/music/index.ts:
```diff
@@ -5,9 +5,45 @@ import { ModuleBase } from '../../internal/module-base';
+ *
+ * For a random album name, use [`albumName()`](https://fakerjs.dev/api/music.html#albumname).
+ *
+ * For a ranom artist, use [`artist()`](https://fakerjs.dev/api/music.html#artist).
```
typo, random

matthewmayer committed:
Merge branch 'next' into music-artist-album-new-genres-songs

jeremyhofer committed:
fix typo. cleanup blatantly profane album names. cleanup non-ascii based names.

matthewmayer commented:
Non ascii names are OK as long as that's how they are normally written in English

ST-DDT commented on src/locales/en/music/album_name.ts:
```diff
@@ -0,0 +1,947 @@
+  '"Awaken, My Love!"',
+  "(What's The Story) Morning Glory?",
+  '- Tragedy +',
+  '.mp3',
```
This might cause confusion and is IMO not worth the confusion.

ST-DDT commented on src/locales/en/music/album_name.ts:
```diff
@@ -0,0 +1,947 @@
+  '- Tragedy +',
+  '.mp3',
+  '13 Reasons Why (Season 3)',
+  '2000',
```
Same here

ST-DDT commented on src/locales/en/music/album_name.ts:
```diff
@@ -0,0 +1,947 @@
+  '2000',
+  '21st Century Breakdown',
+  '30 De Febrero',
+  '42',
```
Same here

ST-DDT commented on src/locales/en/music/album_name.ts:
```diff
@@ -0,0 +1,947 @@
+  '8 Mile',
+  '808s & Heartbreak',
+  '9 To 5 And Odd Jobs',
+  '9',
```
Same here

ST-DDT commented on src/locales/en/music/album_name.ts:
```diff
@@ -0,0 +1,947 @@
+  '808s & Heartbreak',
+  '9 To 5 And Odd Jobs',
+  '9',
+  '99.9%',
```
Same here

ST-DDT commented on src/locales/en/music/album_name.ts:
```diff
@@ -0,0 +1,947 @@
+  '9 To 5 And Odd Jobs',
+  '9',
+  '99.9%',
+  '?',
```
Same here

ST-DDT commented on src/locales/en/music/album_name.ts:
```diff
@@ -0,0 +1,947 @@
+  'Now Or Never',
+  'O Embaixador (Ao Vivo)',
+  'O My Heart',
+  'O',
```
Same here

ST-DDT commented on src/locales/en/music/album_name.ts:
```diff
@@ -0,0 +1,947 @@
+  'Without Warning',
+  'Wonder',
+  'X&Y',
+  'X',
```
Same here

ST-DDT commented on src/locales/en/music/album_name.ts:
```diff
@@ -0,0 +1,947 @@
+  'Youngblood',
+  'Younger Now',
+  'Youth',
+  'xx',
```
I think we should use case insensitive sort

ST-DDT commented on src/locales/en/music/album_name.ts:
```diff
@@ -0,0 +1,947 @@
+export default [
+  '"Awaken, My Love!"',
+  "(What's The Story) Morning Glory?",
```
Does this refer to the thing I think it does?

matthewmayer replied:
It's a double entendre for sure, but it's also a pretty famous album. I don't think it's offensive.

ST-DDT commented on src/locales/en/music/artist.ts:
```diff
@@ -0,0 +1,988 @@
+  'Christopher Cross',
+  'Chuck Berry',
+  'Ciara',
+  'Cigarettes After Sex',
```
Should we remove this in case of value transformation by the user? (aka `value.split(' ')[2]`)
I think we have enough other values to omit this one.

jeremyhofer replied:
I'm not sure about the rationale of removing due to a value transformation. However, I agree there are more than enough examples and have no issue removing this one.

xDivisionByZerox replied:
> Should we remove this in case of value transformation by the user? (aka value.split(' ')[2])

I would not consider this as custom value transformation doesn't make any sense in this context.

> I think we have enough other values to omit this one.

Fair enought. Although, with over 5 Million sold albums world wide they are far more well known than the german rapper CRO (line 182) for example. He has not even 2 Million sold albums.  
Excluding them only because they have an undired word in their name feels kinda wrong to me. While I agree on the fact that we should have somewhat of an profanity filter, I'd not agree on it being applied to names. In the worst case scenario this could potential hurt the overall diversity of values generated, as giving yourself an "outragous" name (as an artist) might a percise decision you make to guid attention to specific topics.

matthewmayer commented:
FWIW although we would normally wait for 10 upvotes to approve a change like this this does seem a fairly sensible extension of the current music module so I'd be fine with approving this for next minor or major release.

jeremyhofer committed:
prune additional artists and albums. case insensitive sort sets

jeremyhofer commented:
Updated to remove albums that could be confusing - the ones that were called out, as well as others I felt could similarly cause some confusion. Updated to case insensitive sort the data as well.

Thanks for considering these methods! Would it be possible to add the ask to propose and vote on new modules and methods to the docs and CONTRIBUTING.md? I was unaware this was the general expectation when I got started in the repo :smiley:

matthewmayer commented:
> Would it be possible to add the ask to propose and vote on new modules and methods to the docs and CONTRIBUTING.md? I was unaware this was the general expectation when I got started in the repo 😃

Its not really a hard requirement, its more to help triage the issues that are added to the project. So if someone adds an issue saying "it would be cool if we had artist names" we will normally wait until >10 upvotes to see if other people would find it useful. If someone sources suitable data, makes a PR, adds tests etc, then the threshold for inclusion is lower!

So I think being too strict about it and documenting in CONTRIBUTING might discourage people from actually contributing code, which we wouldnt want.

xDivisionByZerox commented on src/modules/music/index.ts:
```diff
@@ -5,9 +5,45 @@ import { ModuleBase } from '../../internal/module-base';
+   * @example
+   * faker.music.albumName() // '1989'
+   *
+   * @since 8.5.0
```
We will not be releasing a version 8.5 as we already work on bringing the current codebase to v9.0. So these methods will (most likely) land in v9.0 as well.

```suggestion
   * @since 9.0.0
```

xDivisionByZerox commented on src/modules/music/index.ts:
```diff
@@ -5,9 +5,45 @@ import { ModuleBase } from '../../internal/module-base';
+   * @example
+   * faker.music.artist() // 'The Beatles'
+   *
+   * @since 8.5.0
```
Same here.

```suggestion
   * @since 9.0.0
```

xDivisionByZerox reviewed (changes requested):
Some small updates to the documentation. Otherwise this looks good to me.

ST-DDT committed:
Merge branch 'next' into music-artist-album-new-genres-songs

ST-DDT commented:
We talked about this PR in today's team meeting.
There we discussed whether the methods (and definitions) should use the `Name` suffix.
In our oppinion `album` and `albumName` convey the same information and thus we should only use `album()`.

The definitions should follow the method's name unless there is a need for disambiaguation such as with `_pattern`.
(Since we don't plan to add patterns for these methods we should use the method name for the definitions)

We will rename `songName` -> `song` in a separate PR.

@matthewmayer What do you think?

matthewmayer commented:
Sure. Shorter is better like city instead of cityName

ST-DDT commented:
@jeremyhofer Could you please apply the requested changes and fix the merge conflicts?

xDivisionByZerox committed:
chore: apply suggestions from code review

xDivisionByZerox committed:
Merge branch 'next' into pr/jeremyhofer/2620

xDivisionByZerox committed:
refactor: shorten function names

xDivisionByZerox committed:
docs: fix broken links

ST-DDT commented on src/modules/music/index.ts:
```diff
@@ -5,9 +5,45 @@ import { ModuleBase } from '../../internal/module-base';
- * Generate a random music genre with [`genre()`](https://fakerjs.dev/api/music.html#genre) or song name with [`songName()`](https://fakerjs.dev/api/music.html#songname). Both may be localized.
+ * Generate random music content.
+ *
+ * For a random album name, use [`albumName()`](https://fakerjs.dev/api/music.html#album).
```
```suggestion
 * For a random album name, use [`album()`](https://fakerjs.dev/api/music.html#album).
```

ST-DDT commented on src/locales/en/music/album_name.ts:
The file should probably be renamed to just `album`.
same for song name.

ST-DDT commented on src/definitions/music.ts:
```diff
@@ -4,6 +4,16 @@ import type { LocaleEntry } from './definitions';
+  /**
+   * The names of some albums.
+   */
+  album_name: string[];
```
This should be renamed to just album.
Same for song name.

ST-DDT commented on src/modules/music/index.ts:
```diff
@@ -5,9 +5,45 @@ import { ModuleBase } from '../../internal/module-base';
+ *
+ * For a random genre, use [`genre()`](https://fakerjs.dev/api/music.html#genre).
+ *
+ * For a random song name, [`songName()`](https://fakerjs.dev/api/music.html#song).
```
```suggestion
 * For a random song name, [`song()`](https://fakerjs.dev/api/music.html#song).
```

ST-DDT commented on src/modules/music/index.ts:
```diff
@@ -24,11 +60,11 @@ export class MusicModule extends ModuleBase {
    * @since 7.1.0
    */
-  songName(): string {
+  song(): string {
```
This is a breaking change

xDivisionByZerox replied:
Yeah, I'll revert this. I overred the "in a separate PR" thing in [the original change request comment](https://github.com/faker-js/faker/pull/2620#issuecomment-2050181213).

xDivisionByZerox committed:
docs: fix incorrect method name reference

xDivisionByZerox committed:
refactor: remove name suffix from definitions

xDivisionByZerox committed:
revert: changes to song/songName

xDivisionByZerox committed:
test: update snapshot

xDivisionByZerox committed:
Merge branch 'next' into music-artist-album-new-genres-songs

xDivisionByZerox committed:
refactor: normalize locale data

xDivisionByZerox committed:
test: update snapshots
